pkg/envoy/ads/stream.go

```diff
@@ -136,82 +138,104 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 // an xDS DiscoveryResponse.
 func respondToRequest(proxy *envoy.Proxy, discoveryRequest *xds_discovery.DiscoveryRequest) bool {
 	var err error
 	var requestVersion uint64
 	var requestNonce string
-	var lastVersion uint64
 	var lastNonce string
 
 	log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Request %s [nonce=%s; version=%s; resources=%v] last sent [nonce=%s; version=%d]",
 		proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.TypeUrl,
 		discoveryRequest.ResponseNonce, discoveryRequest.VersionInfo, discoveryRequest.ResourceNames,
 		proxy.GetLastSentNonce(envoy.TypeURI(discoveryRequest.TypeUrl)), proxy.GetLastSentVersion(envoy.TypeURI(discoveryRequest.TypeUrl)))
 
-	if discoveryRequest.ErrorDetail != nil {
-		log.Error().Msgf("Proxy SerialNumber=%s PodUID=%s: [NACK] err: \"%s\" for nonce %s, last version applied on request %s",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ErrorDetail, discoveryRequest.ResponseNonce, discoveryRequest.VersionInfo)
-		return false
-	}
-
+	// Parse TypeURL of the request
 	typeURL, ok := envoy.ValidURI[discoveryRequest.TypeUrl]
 	if !ok {
 		log.Error().Msgf("Proxy SerialNumber=%s PodUID=%s: Unknown/Unsupported URI: %s",
 			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.TypeUrl)
 		return false
 	}
 
-	// It is possible for Envoy to return an empty VersionInfo.
-	// When that's the case - start with 0
-	if discoveryRequest.VersionInfo != "" {
-		if requestVersion, err = strconv.ParseUint(discoveryRequest.VersionInfo, 10, 64); err != nil {
-			// It is probable that Envoy responded with a VersionInfo we did not understand
-			log.Error().Err(err).Msgf("Proxy SerialNumber=%s PodUID=%s: Error parsing DiscoveryRequest with TypeURL=%s VersionInfo=%s (%v)",
-				proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), discoveryRequest.VersionInfo, err)
-			return false
-		}
+	// Parse ACK'd verion on the proxy for this given resource
+	requestVersion, err = parseRequestVersion(discoveryRequest)
+	if err != nil {
+		log.Error().Err(err).Msgf("Proxy SerialNumber=%s PodUID=%s: Error parsing version %s for type %s",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.VersionInfo, typeURL)
+		return false
 	}
 
-	// Set last version applied
-	proxy.SetLastAppliedVersion(typeURL, requestVersion)
+	// Handle NACK case
+	if discoveryRequest.ErrorDetail != nil {
+		log.Error().Msgf("Proxy SerialNumber=%s PodUID=%s: [NACK] err: \"%s\" for nonce %s, last version applied on request %s",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ErrorDetail, discoveryRequest.ResponseNonce, discoveryRequest.VersionInfo)
+		// TODO: if NACK's on our latest nonce, we can also update lastAppliedVersion
+		// TODO: if the NACK's nonce is our latest nonce, we should retry to avoid leaving the envoy in a wrong config state and update
```

Comment:
Are these pending issues tracked in an issue? Would be good to have issues for these.

Reply:
https://github.com/openservicemesh/osm/issues/3254